CONTRIBUTING.md

```diff
@@ -333,6 +333,137 @@ There are several steps that go into a major release
+functioning in a deterministic way.
+
+The state compatibility is ensured by taking the app hash of the state and comparing it
+to the app hash with the rest of the network. Essentially, an app hash is a hash of
+hashes of every store's Merkle root.
+
+For example, at height n, we compute:
+`app hash = hash(hash(root of x/epochs), hash(root of  x/gamm)...)`
+
+Then, Tendermint ensures that the app hash of the local node matches the app hash
+of the network. Please note that the explanation and examples are simplified.
+
+#### Sources of State-incompatibility
+
+##### Creating Additional State
+
+By erroneously creating database entries that exist in Version A but not in
+Version B, we can cause the app hash to differ across nodes running
+these versions in the network. Therefore, this must be avoided.
+
+##### Returning Different Errors Given Same Input
+
+Version A
+```go
+func (sk Keeper) validateAmount(ctx context.Context, amount sdk.Int) error {
+    if amount.IsNegative() {
+        return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "amount must be positive or zero")
+    }
+    return nil, errors.New("error")
+}
+```
+
+Version B
+```go
+func (sk Keeper) validateAmount(ctx context.Context, amount sdk.Int) error {
+    if amount.IsNegative() || amount.IsZero() {
+        return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "amount must be positive")
+    }
+    return nil, errors.New("error")
+}
+```
+
+Note that now an amount of 0 can be valid in "Version A". However, not in "Version B".
+Therefore, if some nodes are running "Version A" and others are running "Version B",
+the final app hash might not be deterministic.
+
+##### Variability in Gas Usage
+
+For transaction flows (or any other flow that consumes gas), it is important
+that the gas usage is deterministic.
+
+Currently, gas usage is being Merklized in the state. As a result, reordering functions
+becomes risky.
+
+Suppose my gas limit is 2000 and 1600 is used up before entering
+`someInternalMethod`. Consider the following:
+
+```go
+func someInternalMethod(ctx sdk.Context) {
+  object1 := readOnlyFunction1(ctx) # consumes 1000 gas
+  object2 := readOnlyFunction2(ctx) # consumes 500 gas
+  doStuff(ctx, object1, object2)
+}
+```
+- It will run out of gas with `gasUsed = 2600` where 2600 getting merkelized
+into the tx results.
+
+```go
+func someInternalMethod(ctx sdk.Context) {
+  object2 := readOnlyFunction2(ctx) # consumes 500 gas
+  object1 := readOnlyFunction1(ctx) # consumes 1000 gas
+  doStuff(ctx, object1, object2)
+}
+```
+- It will run out of gas with `gasUsed = 2100` where 2100 is getting merkelized
+into the tx results.
+
+Therefore, we introduced a state-incompatibility by merklezing diverging gas
+usage.
+
+##### Network Requests
+
+It is critical to avoid performing network requests since it is common
+for services to be unavailable or rate-limit. As a result, nodes
+may get diverging responses, leading to state breakage.
```

Comment:
Hrmm not sure I follow. Maybe we can clarify by stating what is the state machine scope, which is:
* All messages supported by the chain
* Queries in {whitelist file}
* All BeginBlock/Endblock logic

Importantly events are not state machine versioned

Reply:
Will add the suggestions.

This specific part refers to querying external services, for example, some API for getting exchange rates. This is not allowed since such API might be unavailable in some regions, rate limit or just be down.